compiler/rustc_borrowck/src/diagnostics/conflict_errors.rs

```diff
@@ -1252,19 +1254,90 @@ impl<'cx, 'tcx> MirBorrowckCtxt<'cx, 'tcx> {
+                    finder
+                }
+            }
+            impl<'hir> Visitor<'hir> for ExpressionFinder<'hir> {
+                fn visit_expr(&mut self, ex: &'hir hir::Expr<'hir>) {
+                    if ex.span == self.right_span
+                        && let Some(hir::Node::Expr(parent_expr)) = self.hir.find(self.hir.parent_id(ex.hir_id))
+                    {
+                        self.parent_expr = Some(parent_expr);
+                    } else {
+                        hir::intravisit::walk_expr(self, ex);
+                    }
+                }
+            }
+
+            let hir = self.infcx.tcx.hir();
+            let local_ty = self.body.local_decls[place.local].ty;
+
+            // Check if this is a call to `mem::swap` and show a more specialized help message
+            // suggesting `slice::swap`
+            if let Some(hir::Node::Item(item)) = hir.find(self.mir_hir_id())
+                && let hir::ItemKind::Fn(_, _, body_id) = item.kind
+                && let Some(hir::Node::Expr(body)) = hir.find(body_id.hir_id)
+                && let typeck_results = self.infcx.tcx.typeck(self.mir_def_id())
+                && let ExpressionFinder { parent_expr: Some(parent_expr), .. } = ExpressionFinder::find_parent(hir, mut_borrow_span, body)
+
+                && let hir::ExprKind::Call(callee, _) = &parent_expr.kind
+                && let hir::ExprKind::Path(qpath) = &callee.kind
+                && let Res::Def(DefKind::Fn, callee_def_id) = typeck_results.qpath_res(qpath, callee.hir_id)
+                && let Some(mem_swap_def_id) = self.infcx.tcx.get_diagnostic_item(sym::mem_swap)
+                && mem_swap_def_id == callee_def_id
+                && (local_ty.is_slice() || local_ty.is_array())
+
+                && let hir::ExprKind::Call(_, [left_expr, right_expr]) = &parent_expr.kind
+                && let hir::ExprKind::AddrOf(.., left_expr) = &left_expr.kind
+                && let hir::ExprKind::AddrOf(.., right_expr) = &right_expr.kind
+                && let hir::ExprKind::Index(left_expr, left_index) = &left_expr.kind
+                && let hir::ExprKind::Index(_, right_index) = &right_expr.kind
+
+                && let source_map = self.infcx.tcx.sess.source_map()
+                && let Ok(target_snippet) = source_map.span_to_snippet(left_expr.span)
+                && let Ok(left_index_snippet) = source_map.span_to_snippet(left_index.span)
+                && let Ok(right_index_snippet) = source_map.span_to_snippet(right_index.span)
+            {
+                err.span_suggestion_with_style(
+                    parent_expr.span,
+                    "consider using `slice::swap` to swap elements at the specified indices",
+                    format!("{}.swap({}, {})", target_snippet, left_index_snippet, right_index_snippet),
+                    Applicability::MachineApplicable,
+                    SuggestionStyle::ShowAlways
+                );
```

Comment:
I don't think all this code is worth it over just adding a mention of `<[_]>::swap` to the existing diagnostic.

Reply:
Why? Specific machine applicable suggestions are always better than just guesses, no?  It also shows the user exactly how it's meant to be used (though here it's kinda obvious I guess). But fair, if it's just not worth the reviewing/maintaining effort (is there another issue with having this code that is already written anyway?), I'm fine with also just simplfying it to always suggest `slice::swap` and remove all that logic for finding the `mem::swap` expression

Reply:
I agree with @matthewjasper here. This code is really complicated.

Also, I'm not sure if this works if:
* The indexed expressions are pulled out into variables
* The LHS of either index expression is aliased (i.e. `let x = &mut y; swap(&mut x[0], &mut y[1])` I think...

How does it behave on this code?

```rust
fn main() {
    let mut foo = [1, 2, 3, 4];
    let a = &mut foo[1];
    let b = &mut foo[2];
    std::mem::swap(a, b);
}
```

Reply:
okay, it's been a while since I opened this and the more I think about it, I agree. What do we want to do instead? Just mention the existence of the `swap` method on slices like we are doing for `split_at` and get rid of all this visitor stuff and the huge let chain, or scrap this idea entirely? 😅 

As for the code snippet: it ignores it
<details>
<summary>Output</summary>

```
error[E0499]: cannot borrow `foo[_]` as mutable more than once at a time
  --> test.rs:31:13
   |
30 |     let a = &mut foo[1];
   |             ----------- first mutable borrow occurs here
31 |     let b = &mut foo[2];
   |             ^^^^^^^^^^^ second mutable borrow occurs here
32 |     std::mem::swap(a, b);
   |                    - first borrow later used here
   |
   = help: consider using `.split_at_mut(position)` or similar method to obtain two mutable non-overlapping sub-slices

error: aborting due to previous error
```
</details>

Reply:
Yeah, let's just mention the existence of the method in a note.